packages/react-devtools-shared/src/devtools/views/Profiler/HookChangeSummary.js

```diff
@@ -0,0 +1,185 @@
+import * as React from 'react';
+import {
+  useContext,
+  useMemo,
+  useCallback,
+  memo,
+  useState,
+  useEffect,
+} from 'react';
+import styles from './HookChangeSummary.css';
+import ButtonIcon from '../ButtonIcon';
+import {InspectedElementContext} from '../Components/InspectedElementContext';
+import {StoreContext} from '../context';
+import {
+  HooksList,
+  State,
+} from 'react-devtools-shared/src/devtools/views/Profiler/types';
+
+import {
+  getAlreadyLoadedHookNames,
+  getHookSourceLocationKey,
+} from 'react-devtools-shared/src/hookNamesCache';
+import Toggle from '../Toggle';
+
+const hookListFormatter = new Intl.ListFormat('en', {
+  style: 'long',
+  type: 'conjunction',
+});
+
+const Hook = memo(({hook, hookNames}) => {
+  const hookSource = hook.hookSource;
+  const hookName = useMemo(() => {
+    if (!hookSource || !hookNames) return null;
+    const key = getHookSourceLocationKey(hookSource);
+    return hookNames.get(key) || null;
+  }, [hookSource, hookNames]);
+
+  return (
+    <ul className={styles.Hook}>
+      <li>
+        {hook.id !== null && (
+          <span className={styles.PrimitiveHookNumber}>{hook.id + 1}</span>
+        )}
+        <span
+          className={hook.id !== null ? styles.PrimitiveHookName : styles.Name}>
+          {hook.name}
+          {hookName && <span className={styles.HookName}>({hookName})</span>}
+        </span>
+        {hook.subHooks?.map((subHook, index) => (
+          <Hook
+            key={`${hook.id}-${index}`}
+            hook={subHook}
+            hookNames={hookNames}
+          />
+        ))}
+      </li>
+    </ul>
+  );
+});
+
+const shouldKeepHook = (hook, hooksArray) => {
+  if (hook.id !== null && hooksArray.includes(hook.id)) {
+    return true;
+  }
+  return (
+    hook.subHooks?.some(subHook => shouldKeepHook(subHook, hooksArray)) ?? false
+  );
+};
+
+const filterHooks = (hook, hooksArray) => {
+  if (!shouldKeepHook(hook, hooksArray)) {
+    return null;
+  }
+
+  if (hook.subHooks?.length > 0) {
+    const filteredSubHooks = hook.subHooks
+      .map(subHook => filterHooks(subHook, hooksArray))
+      .filter(Boolean);
+
+    return filteredSubHooks.length > 0
+      ? {...hook, subHooks: filteredSubHooks}
+      : {...hook};
+  }
+
+  return hook;
+};
+
+type Props = {
+  fiberID: number,
+  hooks: HooksList | null,
+  state: State | null,
+  displayMode?: 'detailed' | 'compact',
+};
+const HookChangeSummary = memo(
+  ({hooks, fiberID, state, displayMode = 'detailed'}: Props) => {
+    const {parseHookNames, toggleParseHookNames, inspectedElement} = useContext(
+      InspectedElementContext,
+    );
+    const store = useContext(StoreContext);
+
+    const [parseHookNamesOptimistic, setParseHookNamesOptimistic] =
+      useState(parseHookNames);
+
+    useEffect(() => {
+      setParseHookNamesOptimistic(parseHookNames);
+    }, [inspectedElement?.id, parseHookNames]);
```

Comment:
This looks strange, you are keeping something from a global context in a local state, and then reconciliate it here? Why `parseHookNames` from `InspectedElementContext` is not enough?

Reply:
This approach is something I borrowed from the Inspected Element Panel [here](https://github.com/facebook/react/blob/79ddf5b574db085104d917c24a964cbd5b824e09/packages/react-devtools-shared/src/devtools/views/Components/InspectedElementHooksTree.js#L57).

The idea behind the `parseHookNamesOptimistic` state is to keep the UI responsive. Since changing `parseHookNames` happens in a transition (which can suspend), this local state ensures the toggle feels immediate while the actual state updates in the background.

I considered making this logic reusable, by extracting `LoadHookNamesButton` into its own component, but I didn’t want to optimize prematurely. Do you think it’s worth revisiting?

Reply:
Oh, okay, we should probably refactor this to use `useOptimistic` in the future.